core/store/src/trie/prefetching_trie_storage.rs

```diff
@@ -0,0 +1,417 @@
+    /// pre-fetcher uses this in read-only mode to avoid premature evictions.
+    shard_cache: TrieCache,
+    /// Shared with parent `TrieCachingStorage`.
+    prefetching: Arc<Mutex<PrefetchStagingArea>>,
+}
+
+/// This type is shared between runtime crate and store crate.
+///
+/// The former creates `PrefetchApi` and puts requests in, the latter serves requests.
+/// With this API, the store does not know about receipts etc, and the runtime
+/// does not know about the trie structure. The only thing they share is this object.
+#[derive(Clone)]
+pub struct PrefetchApi {
+    /// Bounded, shared queue for all IO threads to take work from.
+    ///
+    /// Work items are defined as `TrieKey` because currently the only
+    /// work is to prefetch a trie key. If other IO work is added, consider
+    /// changing the queue to an enum.
+    work_queue: Arc<Mutex<BoundedQueue<TrieKey>>>,
+    /// Prefetching IO threads will insert fetched data here. This is also used
+    /// to mark what is already being fetched, to avoid fetching the same data
+    /// multiple times.
+    prefetching: Arc<Mutex<PrefetchStagingArea>>,
+    /// Set to true to stop all io threads.
+    stop_io: Arc<AtomicBool>,
+}
+
+/// Staging area for in-flight prefetch requests and a buffer for prefetched data.
+///
+/// Before starting a pre-fetch, a slot is reserved for it. Once the data is
+/// here, it will be put in that slot. The parent `TrieCachingStorage` needs
+/// to take it out and move it to the shard cache.
+///
+/// A shared staging area is the interface between `TrieCachingStorage` and
+/// `TriePrefetchingStorage`. The parent simply checks the staging area before
+/// going to the DB. Otherwise, no communication between the two is necessary.
+///
+/// This design also ensures the shard cache works exactly the same with or
+/// without the prefetcher, because the order in which it sees accesses is
+/// independent of the prefetcher.
+#[derive(Default)]
+pub(crate) struct PrefetchStagingArea {
+    slots: HashMap<CryptoHash, PrefetchSlot>,
+    size_bytes: usize,
+}
+
+/// Result when atomically accessing the prefetch staging area.
+pub(crate) enum PrefetcherResult {
+    SlotReserved,
+    Pending,
+    Prefetched(Arc<[u8]>),
+    MemoryLimitReached,
+}
+
+/// Type used interanlly in the staging area to keep track of requests.
+#[derive(Clone, Debug)]
+enum PrefetchSlot {
+    PendingPrefetch,
+    PendingFetch,
+    Done(Arc<[u8]>),
+}
+
+impl TrieStorage for TriePrefetchingStorage {
+    // Note: This is the tricky bit of the implementation.
+    // We have to retrieve data only once in many threads, so all IO threads
+    // have to go though the staging area and check for inflight requests.
+    // The shard cache mutex plus the prefetch staging area mutex are used for
+    // that in combination. Let's call the first lock S and the second P.
+    // The rules for S and P are:
+    // 1. To avoid deadlocks, S must always be requested before P, if they are
```

Comment:
I'm not sure how step 1 helps here. I think I applied it roughly like you suggested. But the shard cache still lives next to the staging area in `TriePrefetchingStorage` and in `TrieCachingStorage`. Both could still lock in the wrong order from a type-system point of view.
Did I misunderstand step 1, or am I missing something else? It would be lovely if this could be incorporated in the type system!

Reply:
My thinking was that if the mutex is inside PrefetchStagingArea, then it is the only thing that will ever lock it, and by scoping it cannot hold the lock for long enough to cause a deadlock. That said I haven't re-reviewed the PR yet so I may come back in a bit with additional details